### PR TITLE
Initial fixes for HeartbeatService

### PR DESCRIPTION
**Pull request by SerialVelocity**
- [x] Get rid of atomic integer
 - [x] Retry cas multiple times to drop lock
 - [x] Check for lock every 5 secs (or backoff, etc), throw if the lock doesn't heartbeat in 60 secs

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/1035)

### CODE REVIEW

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/HeartbeatService.java [hunk 1]
```diff
@@ -61,7 +59,7 @@
         this.heartbeatExecutorService = null;
     }
 
-    void startBeatingForLock(long lockId) {
+    public synchronized void startBeatingForLock(long lockId) {
```
Comment (jboreiko) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/HeartbeatService.java]: Seems like we should never have multiple threads accessing this class.  Only one thread should ever be able to progress past waitForSchemaMutationLock().  Am I missing something here?

**Comment by ashrayjain** [comment]
@SerialVelocity let me know if you think there is better way to implement some of the stuff. happy to update as required.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/CassandraKeyValueServiceConfig.java [hunk 2]
```diff
@@ -144,11 +144,6 @@ public int schemaMutationTimeoutMillis() {
     }
 
     @Value.Default
-    public int heartbeatTimePeriodMillis() {
```
Comment (gbrova) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/CassandraKeyValueServiceConfig.java]: I noticed this config wasn't documented; we should add a section explaining the heartbeat (and also the schema mutation lock more generally)

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/SchemaMutationLock.java [hunk 3]
```diff
@@ -87,19 +101,22 @@ void runWithLock(Action action) {
         }
 
         long lockId = waitForSchemaMutationLock();
-        heartbeatService.startBeatingForLock(lockId);
         try {
-            action.execute();
-        } catch (Exception e) {
-            throw Throwables.throwUncheckedException(e);
+            try {
```
Comment (gbrova) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/SchemaMutationLock.java]: Not a fan of the nested try/catch; consider extracting a method.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/SchemaMutationLock.java [hunk 4]
```diff
@@ -141,15 +158,16 @@ private long waitForSchemaMutationLock() {
...
                 Column ourUpdate = lockColumnFromIdAndHeartbeat(perOperationNodeId, 0);
 
                 List<Column> expected = ImmutableList.of(lockColumnWithValue(GLOBAL_DDL_LOCK_CLEARED_VALUE));
 
                 CASResult casResult = writeDdlLockWithCas(client, expected, ourUpdate);
 
                 Column lastSeenColumn = null;
-                int timesAttempted = 0;
+                long lastSeenColumnUpdateTs = 0;
+                int currentTimeBetweenLockAttemptsMillis = TIME_BETWEEN_LOCK_ATTEMPT_ROUNDS_MILLIS;
```
Comment (gbrova) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/SchemaMutationLock.java]: Maybe `INITIAL_TIME_BETWEEN_LOCK_ATTEMPT_ROUNDS_MILLIS`?  It's confusing for this to be a final when the actual time between attempts changes.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/SchemaMutationLock.java [hunk 5]
```diff
@@ -253,31 +276,62 @@ private void waitForSchemaMutationLockWithoutCas() throws TimeoutException {
...
+        boolean unlockDone = false;
+        boolean isInterrupted = false;
+        long currentSleepMillis = TIME_BETWEEN_LOCK_ATTEMPT_ROUNDS_MILLIS;
+        for (int unlockRetryCount = 0; unlockRetryCount < MAX_UNLOCK_RETRY_COUNT; unlockRetryCount++) {
+            unlockDone = trySchemaMutationUnlockOnce(perOperationNodeId);
+            if (unlockDone) {
+                break;
+            }
+
+            currentSleepMillis = Math.min(2 * currentSleepMillis, TIME_BETWEEN_LOCK_ATTEMPT_ROUNDS_MILLIS_CAP);
```
Comment (gbrova) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/SchemaMutationLock.java]: This duplicates the Math.min(...) logic from line 209 above; consider extracting a method.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/SchemaMutationLock.java [hunk 6]
```diff
@@ -253,31 +276,62 @@ private void waitForSchemaMutationLockWithoutCas() throws TimeoutException {
...
+            try {
+                Thread.sleep(currentSleepMillis);
+            } catch (InterruptedException e) {
+                isInterrupted = true;
```
Comment (gbrova) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/SchemaMutationLock.java]: Why not just call `Thread.currentThread().interrupt();` here?

Comment (ashrayjain) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/SchemaMutationLock.java]: Calling `interrupt()` sets the interrupt flag. I believe this will cause functions like `sleep` to detect that the thread has been interrupted and throw InterruptedExceptions. We wanted to continue trying even if we get interrupted here. Calling it here will make `sleep` exit in the next iteration (and every other iteration as long as the flag is set).

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/CassandraKeyValueServiceConfig.java (hunk 2 above)
Comment (ashrayjain) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/CassandraKeyValueServiceConfig.java]: We added an ADR for heartbeat with the last PR. Do you feel we need something else added to that?

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/SchemaMutationLock.java (hunk 6 above)
Comment (gbrova) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/SchemaMutationLock.java]: That makes sense. 

I'm not convinced that ignoring the interrupt is best practice (instead of, for example, throwing with something like "Unable to unlock because we were interrupted" or even retrying without the wait as a fast, last-ditch retry attempt).  To be fair, I don't have a clear notion of why an interrupt might occur here - so if you've thought this through and are convinced the current approach is the right way to go, I'm fine for this to merge as is.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/CassandraKeyValueServiceConfig.java (hunk 2 above)
Comment (gbrova) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/CassandraKeyValueServiceConfig.java]: I saw the ADR and think it's well-written. It might be useful to additionally have some user-facing docs, maybe with a heading like "why we have a C* schema mutation lock and what to do if it stops heartbeating". A lot of the content might even be copy-pasted from the ADR, it's just that the ADRs aren't published as part of docs and are therefore harder for users to find / link to. 

Also tagging @rhero for context/input about how we should do docs. 

Having said all this, I want this bugfix PR to merge soon, and wouldn't block on the docs change.

Comment (ashrayjain) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/CassandraKeyValueServiceConfig.java]: Agreed. I think this is a really good idea since it will help educating our users on why we have these things. Also agree that this is probably best done as a separate PR.

**Comment by gbrova** [comment]
Talked to @SerialVelocity, I might be persuaded to leave it as is in order
to minimize the likelihood of dropping the lock (and tolerate not
responding to interrupts as fast)

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/SchemaMutationLock.java (hunk 6 above)
Comment (SerialVelocity) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/SchemaMutationLock.java]: See @gbrova's comment (https://github.com/palantir/atlasdb/pull/1035#issuecomment-252634779).

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/CassandraKeyValueServiceConfig.java (hunk 2 above)
Comment (jboreiko) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/CassandraKeyValueServiceConfig.java]: Good practice he might just be to have a docs section for the adrs, I imagine there is some way to hook them into the docs generation
